language/stdlib/modules/LibraTransaction.move

```diff
@@ -0,0 +1,213 @@
+        );
+
+        prologue_common<Token>(
+            sender,
+            txn_sequence_number,
+            txn_public_key,
+            txn_gas_price,
+            txn_max_gas_units,
+            txn_expiration_time,
+            chain_id,
+        )
+    }
+
+    /// The prologue for script transaction
+    fun script_prologue<Token>(
+        sender: &signer,
+        txn_sequence_number: u64,
+        txn_public_key: vector<u8>,
+        txn_gas_price: u64,
+        txn_max_gas_units: u64,
+        txn_expiration_time: u64,
+        chain_id: u8,
+        script_hash: vector<u8>,
+    ) {
+        assert(
+            LibraTransactionPublishingOption::is_script_allowed(sender, &script_hash),
+            PROLOGUE_ESCRIPT_NOT_ALLOWED
+        );
+
+        prologue_common<Token>(
+            sender,
+            txn_sequence_number,
+            txn_public_key,
+            txn_gas_price,
+            txn_max_gas_units,
+            txn_expiration_time,
+            chain_id,
+        )
+    }
+
+    /// The common prologue is invoked at the beginning of every transaction
+    /// It verifies:
+    /// - The account's auth key matches the transaction's public key
+    /// - That the account has enough balance to pay for all of the gas
+    /// - That the sequence number matches the transaction's sequence key
+    fun prologue_common<Token>(
+        sender: &signer,
+        txn_sequence_number: u64,
+        txn_public_key: vector<u8>,
+        txn_gas_price: u64,
+        txn_max_gas_units: u64,
+        txn_expiration_time_seconds: u64,
+        chain_id: u8,
+    ) {
+        // Check that the chain ID stored on-chain matches the chain ID specified by the transaction
+        assert(ChainId::get() == chain_id, PROLOGUE_EBAD_CHAIN_ID);
+
+        let max_transaction_fee = txn_gas_price * txn_max_gas_units;
+        LibraAccount::check_sender<Token>(sender, max_transaction_fee, txn_sequence_number, txn_public_key);
+
+        assert(
+            LibraTimestamp::now_seconds() < txn_expiration_time_seconds,
+            PROLOGUE_ETRANSACTION_EXPIRED
+        );
+    }
+
+    fun writeset_prologue(
+        account: &signer,
+        writeset_sequence_number: u64,
+        writeset_public_key: vector<u8>,
+    ) {
+        assert(sender == CoreAddresses::LIBRA_ROOT_ADDRESS(), PROLOGUE_EINVALID_WRITESET_SENDER);
+
+        // Currency code don't matter here as it won't be charged anyway.
+        LibraAccount::check_sender<LBR::LBR>(account, 0, writeset_sequence_number, writeset_public_key);
+        // TODO: Check ChainID & Expiration Time?
+    }
+
+    /// The success_epilogue is invoked at the end of successfully executed transactions.
+    fun success_epilogue<Token>(
+        account: &signer,
+        txn_sequence_number: u64,
+        txn_gas_price: u64,
+        txn_max_gas_units: u64,
+        gas_units_remaining: u64
+    ) {
+        let sender = Signer::address_of(account);
+
+        // Charge for gas
```

Comment:
I find this comment confusing given that the actual charging happens in `LibraAccount`. This part is just sanity-checking the gas price and gas unit values.